petab/v2/petab1to2.py

```diff
@@ -0,0 +1,147 @@
+"""Convert PEtab version 1 problems to version 2."""
+import logging
+import shutil
+from itertools import chain
+from pathlib import Path
+
+from pandas.io.common import get_handle, is_url
+
+import petab.C
+from petab.models import MODEL_TYPE_SBML
+from petab.v2.lint import lint_problem as lint_v2_problem
+from petab.yaml import get_path_prefix
+
+from .. import lint_problem
+from ..yaml import load_yaml, validate, write_yaml
+
+logging.basicConfig(level=logging.DEBUG)
+
+
+def petab1to2(yaml_config: Path | str, output_dir: Path | str = None):
+    """Convert from PEtab 1.0 to PEtab 2.0 format.
+
+    Convert a PEtab problem from PEtab 1.0 to PEtab 2.0 format.
+
+    Parameters
+    ----------
+    yaml_config: dict | Path | str
+        The PEtab problem as dictionary or YAML file name.
+    output_dir: Path | str
+        The output directory to save the converted PEtab problem, or ``None``,
+        to return a :class:`petab.v2.Problem` instance.
+
+    Raises
+    ------
+    ValueError
+        If the input is invalid or does not pass linting or if the generated
+        files do not pass linting.
+    """
+    if output_dir is None:
+        # TODO requires petab.v2.Problem
+        raise NotImplementedError("Not implemented yet.")
+    elif isinstance(yaml_config, dict):
+        raise ValueError("If output_dir is given, yaml_config must be a file.")
+
+    if isinstance(yaml_config, Path | str):
+        yaml_file = str(yaml_config)
+        path_prefix = get_path_prefix(yaml_file)
+        yaml_config = load_yaml(yaml_config)
+        get_src_path = lambda filename: f"{path_prefix}/{filename}"  # noqa: E731
+    else:
+        yaml_file = None
+        path_prefix = None
+        get_src_path = lambda filename: filename  # noqa: E731
+
+    get_dest_path = lambda filename: f"{output_dir}/{filename}"  # noqa: E731
```

Comment:
Keeping file names the same when upgrading is nice, but this is more intuitive to me. Fine as is, if you think this is nicer for most users.

```python
problem1 = petab.v1.Problem.get_problem(yaml_config, validate=True)

experiment_df, measurement_df = convert_measurement_df_1to2(problem1)
problem2 = petab.v2.Problem(
    problem1.condition_df,
    experiment_df,
    measurement_df,
    ...,
)

problem2.validate(raise=True)

if output_dir:
    # alternatively `to_files` if you want to keep file names consistent...
    problem2.to_files_generic(output_dir)

return problem2
```

Reply:
The thing is, if you had multiple measurement files, condition files, ... then you most likely still want to have them as separate files in your v2 problem. That's not possible anymore once you read them into a v1 Problem, they will all get merged. And I don't want to rename all files afterwards.

Reply:
Makes sense! I have thought elsewhere that tracking a file ID for e.g. measurements, as an index subset, could improve things a bit. If you're not against this, I can open an issue for it.

Reply:
Also happy to add that to this PR, today or tomorrow.

Reply:
I'm generally open to tracking that information somewhere. To be discussed how that could look in details.

Reply:
> Also happy to add that to this PR, today or tomorrow.

I'd rather have it done separately.